src/zarr/api/asynchronous.py

```diff
@@ -324,7 +325,11 @@ async def open(
         Return type depends on what exists in the given store.
     """
     zarr_format = _handle_zarr_version_or_format(zarr_version=zarr_version, zarr_format=zarr_format)
-
+    if mode is None:
+        if isinstance(store, Store) and store.read_only:
```

Comment:
This should also check for `StorePath`, and if it is a `StorePath` extract the `Store` and follow the same logic.